### DISCUSSION TITLE
ConnectionString set in AppHost is discarded when adding Azure AppConfiguration

### DISCUSSION THREAD

**Discussion by tobey74** [microsoft/aspire#5492] [Q&A] [answered]
Hi, I am using the aspire 'DatabaseContainers' sample but simplified by only spinning up the SQL Server database. 

When AppHost startup everything, the correct ConnectionString pointing to the Docker Desktop container is injected into the ApiService project as expected:

![image](https://github.com/user-attachments/assets/a0821f46-e05b-4ea9-9478-c0cf638ca36e)

But if I connect to our Azure AppConfiguration in the ApiService, then whatever is injected in the AppHost project gets overridden by Azure AppConfiguration.

So calling an enpoint like

```c#
app.MapGet("/connectionstring", () => Results.Ok(app.Configuration.GetConnectionString("Pida") ?? "No ConnectionString called Pida found anywhere!"));
``` 

returns this ConnectionString coming from Azure AppConfiguration:

`"Data Source=localhost;Initial Catalog=sql-db-pida-local;Integrated Security=True;Encrypt=false"`

Instead of the one injected to ApiService by the Aspire AppHost:

`Server=127.0.0.1,56518;User ID=sa;Password=udqkJTz{JpQ.YmetVn2)2E;TrustServerCertificate=true;Database=Pida`

The placement of `.AddAzureAppConfiguration` does not matter. 

```c#
using Azure.Core;
using Azure.Extensions.AspNetCore.Configuration.Secrets;
using Azure.Identity;
using Azure.Security.KeyVault.Secrets;
using DatabaseContainers.ApiService;
using Microsoft.Extensions.Configuration.AzureAppConfiguration;

var builder = WebApplication.CreateBuilder(args);

string appConfig = builder.Configuration.GetValue<string>("AppConfig") ??
    throw new InvalidOperationException("AppConfig uri is missing from configuration/environment");
string uriString = builder.Configuration.GetValue<string>("KeyVault") ??
    throw new InvalidOperationException("KeyVault uri is missing from configuration/environment");

builder.Configuration.AddAzureAppConfiguration((Action<AzureAppConfigurationOptions>)(options => options.Connect(new Uri(appConfig), (TokenCredential)new DefaultAzureCredential()).Select("*").Select("*", "local").ConfigureRefresh((Action<AzureAppConfigurationRefreshOptions>)(refreshOptions => refreshOptions.Register("Sentinel", "\0", true).SetCacheExpiration(TimeSpan.FromSeconds(15.0)))))).AddAzureKeyVault(new SecretClient(new Uri(uriString), (TokenCredential)new DefaultAzureCredential()), new KeyVaultSecretManager());

// Add service defaults & Aspire components.
builder.AddServiceDefaults();

builder.AddSqlServerClient("AddressBook");

// Add services to the container.
builder.Services.AddProblemDetails();
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();

app.UseExceptionHandler();

app.UseSwagger();
if (app.Environment.IsDevelopment())
{
    app.UseSwaggerUI();
}

app.MapAddressBookApi();
app.MapDefaultEndpoints();

app.Run();
``` 

I was under the impression that we can easily add .NET Aspire orchestration to our existing apps during development, and by simply injecting the right environment variables through the AppHost the app could then use a SQL Server container in Docker Desktop, but later when app is deployed to Kubernetes without the AppHost project, our app will simply get the "Pida" ConnectionString from Azure AppConfiguration.

**Comment by davidfowl** [comment]
This is how the .NET configuration system works. Providers are ordered and you’re adding app configuration last. Aspire isn’t doing anything magical here.

**Reply by tobey74** [reply]
Thanks for pointing that out @davidfowl. I was just under the impression that when starting the app stack through the AppHost project and running as an Aspire orchestration, then Aspire would be the last config provider added?

Obviously I am wrong though, but then how can you add Aspire orchestration to an existing app that currently depends upon Azure AppConfiguration values, so that when doing local development values injected from the AppHost are used? But calls for keys not injected by the AppHost will still be retrieved from Azure AppConfig or any appsettings.json files present of course.

Basically ensuring that Aspire/AppHost is added last as an app configuration provider?

**Reply by davidfowl** [reply, marked as answer]
>  then Aspire would be the last config provider added?

There's no aspire configuration provider. The apphost is setting environment variables which is the second last provider added by default.

>Basically ensuring that Aspire/AppHost is added last as an app configuration provider?

By adding AddEnvironmentVariables after adding the app configuration based configuration.

```C#
builder.Configuration.AddEnvironmentVariables();
```
